tests/PyomoTest.py

```diff
@@ -0,0 +1,283 @@
+  ## - source ("generator")
+  ##   - has a capex, fixed OM (yearly recurring), and variable OM (hourly recurring)
+  ##   - variable total size
+  ##   - variable dispatch
+  ## - sink ("market")
+  ##   - has sales profit (hourly recurring)
+  ##   - fixed size
+
+  # then we start creating components
+  tealComps = {}
+
+
+  # *** GENERATOR ***
+  # make the Pyomo variables, constants
+  generator_life = 5 # years of operation before replacement
+  generator_size = pyo.Var(initialize=1, bounds=(0, 100)) # build size, limit to 100 MW
+  m.gen_size = generator_size
+
+  # set up the generator's dispatch variables
+  generator_dispatch = np.zeros((generator_life, hours_in_year), dtype=object)
+  for y in range(generator_life):
+    # make a new pyomo var for this "year"'s dispatch
+    var = pyo.Var(list(range(hours_in_year)), initialize=lambda m, t: 0, bounds=lambda m, t: (0, 100)) #generator_size))
+    setattr(m, f'Gen_disp_year_{y+1}', var)
+    generator_dispatch[y, :] = np.array(list(var.values()))
+
+  generator = build_generator(generator_size, generator_life, generator_dispatch)
+
+  # *** MARKET ***
+  # make the TEAL component
+  market_size = 50 # market depth
+  market_life = 5 # years of operation before replacement
+  # market prices should be interesting, say cos(x) + y/10 where x is hour and y is year
+  prices = np.ones((market_life, hours_in_year))
+  prices[:] *= np.cos(hours) * 100
+  for y in range(market_life):
+    if y % 2 == 0:
+      prices[y, :] *= -1
+
+  # market has similar dispatch to generator, that we'll attach through constraints
+  market_dispatch = np.zeros((market_life, hours_in_year), dtype=object)
+  for y in range(market_life):
+    # make a new pyomo var for this "year"'s dispatch
+    var = pyo.Var(list(range(hours_in_year)), initialize=lambda m, t: 0, bounds=lambda m, t: (0, market_size))
+    setattr(m, f'Market_disp_year_{y+1}', var)
+    market_dispatch[y, :] = np.array(list(var.values()))
+
+  market = build_market(market_size, market_life, prices, market_dispatch)
+
+  # DEBUGG
+
+  metrics = RunCashFlow.run(tealSettings, [generator, market], {}, pyomoVar=True) # Past version was pyomoChk
+
+  # now set up the rest of the Pyomo model
+
+  # *** Constraints ***
+  # hourly conservation of electricity
+  def conserve_elec(y, m, t):
+    gen = getattr(m, f'Gen_disp_year_{y+1}')
+    mkt = getattr(m, f'Market_disp_year_{y+1}')
+    return gen[t] == mkt[t]
+
+  for y in range(project_life):
+    con = pyo.Constraint(m.T, rule=partial(conserve_elec, y))
+    setattr(m, f'conserve_{y+1}', con)
+
+  # hourly dispatch < gen size
+  def size_limit(size, y, m, t):
+    gen = getattr(m, f'Gen_disp_year_{y+1}')
+    return gen[t] <= size
+
+  for y in range(project_life):
+    lim = pyo.Constraint(m.T, rule=partial(size_limit, generator_size, y))
+    setattr(m, f'size_{y+1}', lim)
+
+  # Objective: NPV
+  m.npv = pyo.Objective(expr=metrics['NPV'], sense=pyo.maximize)
+
+  print('DEBUGG PRE:')
```

Comment:
how much of this do we want printing?

Reply:
We don't really need the pre-printing but post-printing I would like to keep